relay/pkg/block/pull_test.go

```diff
@@ -0,0 +1,261 @@
+	}
+
+	headers := []*btc.Header{
+		{Hash: [32]byte{1}, Height: 1, PrevHash: [32]byte{0}},
+		{Hash: [32]byte{2}, Height: 2, PrevHash: [32]byte{1}},
+	}
+
+	for _, header := range headers {
+		forwarder.pushHeaderToQueue(header)
+	}
+
+	// Check nextPullHeaderHeight
+	expectedNextPullHeight := int64(3)
+	actualNextPullHeight := forwarder.nextPullHeaderHeight
+
+	if actualNextPullHeight != expectedNextPullHeight {
+		t.Errorf(
+			"unexpected add headers event:\n"+
+				"expected: [%d]\n"+
+				"actual:   [%d]\n",
+			expectedNextPullHeight,
+			actualNextPullHeight,
+		)
+	}
+
+	// Check headers on the queue
+	expectedNoOfHeaders := 2
+	if len(forwarder.headersQueue) != expectedNoOfHeaders {
+		t.Errorf(
+			"unexpected number of headers in channel:\n"+
+				"expected: [%d]\n"+
+				"actual:   [%d]\n",
+			expectedNoOfHeaders,
+			len(forwarder.headersQueue),
+		)
+	}
+
+	for _, expectedHeader := range headers {
+		actualHeader := <-forwarder.headersQueue
+		if !actualHeader.Equals(expectedHeader) {
+			t.Errorf(
+				"unexpected header in queue:\n"+
+					"expected: [%s]\n"+
+					"actual:   [%s]\n",
+				expectedHeader.String(),
+				actualHeader.String(),
+			)
+		}
+	}
+}
+
+func TestPullHeaderFromBtcChain(t *testing.T) {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	defer cancelCtx()
+
+	bc, err := btclocal.Connect()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	btcChain := bc.(*btclocal.Chain)
+
+	btcChain.SetHeaders([]*btc.Header{
+		{Height: 1, Hash: [32]byte{1}, Raw: []byte{1}},
+		{Height: 2, Hash: [32]byte{2}, Raw: []byte{2}},
+	})
+
+	forwarder := &Forwarder{
+		btcChain:                  btcChain,
+		forwarderPullingSleepTime: 3 * time.Second,
+		nextPullHeaderHeight:      1,
+	}
+
+	// Test that we can pull headers without waiting when we have not reached
+	// Bitcoin chain tip
+	var headers []btc.Header
+	for i := 0; i < 2; i++ {
+		header, err := forwarder.pullHeaderFromBtcChain(ctx)
+		if err != nil {
+			t.Fatal(err)
+		}
+		headers = append(headers, *header)
+		forwarder.nextPullHeaderHeight++
+	}
+
+	expectedHeaders := []btc.Header{
+		{Height: 1, Hash: [32]byte{1}, Raw: []byte{1}},
+		{Height: 2, Hash: [32]byte{2}, Raw: []byte{2}},
+	}
+
+	if !reflect.DeepEqual(expectedHeaders, headers) {
+		t.Errorf(
+			"unexpected headers:\n"+
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedHeaders,
+			headers,
+		)
+	}
+
+	// Test that function hangs after we have reached the tip of Bitcoin chain
+	// and waits for more blocks to be appended
+	headerChan := make(chan *btc.Header, 1)
+
+	go func() {
+		header, err := forwarder.pullHeaderFromBtcChain(ctx)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		headerChan <- header
+	}()
+
+	select {
+	case <-time.After(time.Second):
+	case header := <-headerChan:
+		t.Fatalf("unexpected header returned[%v]", header)
+	}
+
+	btcChain.AppendHeader(
+		&btc.Header{Height: 3, Hash: [32]byte{3}, Raw: []byte{3}},
+	)
+
+	select {
+	case <-time.After(3 * time.Second):
```

Comment:
If we shorten the loop sleeping time, we can use a smaller value here as well.

Reply:
Shortened it to 300ms.